src/sentry/nodestore/base.py

```diff
@@ -85,3 +94,46 @@ def generate_id(self):
 
     def cleanup(self, cutoff_timestamp):
         raise NotImplementedError
+
+    def _get_cache_item(self, id):
+        if self.cache and self.sample_rate != 0.0:
+            return self.cache.get(id)
+
+    def _get_cache_items(self, id_list):
+        if self.cache and self.sample_rate != 0.0:
+            return self.cache.get_many(id_list)
+        return {}
+
+    def _set_cache_item(self, id, data):
+        if self.cache and data:
+            if self.should_cache(id):
+                self.cache.set(id, data)
+
+    def _set_cache_items(self, items):
+        cacheable_items = {k: v for k, v in six.iteritems(items) if v and self.should_cache(k)}
+        if self.cache:
+            self.cache.set_many(cacheable_items)
+
+    def _delete_cache_item(self, id):
+        if self.cache:
+            self.cache.delete(id)
+
+    def _delete_cache_items(self, id_list):
+        if self.cache:
+            self.cache.delete_many(id_list)
+
+    @memoize
+    def cache(self):
+        try:
+            return caches["nodedata"]
+        except InvalidCacheBackendError:
+            return None
+
+    @memoize
+    def sample_rate(self):
+        from sentry import options
+
+        return options.get("nodedata.cache-sample-rate", 0.0)
+
+    def should_cache(self, id):
+        return (int(md5(id).hexdigest(), 16) / float(int("f" * 32, 16))) < self.sample_rate
```

Comment:
I think you can obtain the same result in a more robust way

> ((int(md5(id).hexdigest(), 16) % 1000) / 1000.0

Also this breaks on python3. please keep it in mind since in python 3 the string will be unicode and will need to be encoded before passing it to md5

Reply:
That's a lot better, thanks